lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -535,15 +535,26 @@ newDBLayerWith _cacheBehavior tr ti SqliteContext{runQuery} = do
     -- mutex (queryLock), which means no concurrent queries.
     queryLock <- newMVar () -- fixme: ADP-586
 
-    -- Insert a checkpoint into the DBVar
+    -- Insert a checkpoint for a given wallet ID into the DBVar,
+    -- assuming that the wallet already exists.
     let insertCheckpointCached wid cp = do
             liftIO $ traceWith tr $ MsgCheckpointCache wid MsgPutCheckpoint
             modifyDBMaybe checkpointsVar $ \ws ->
                 let point = getPoint cp
                 in  case Map.lookup wid ws of
-                    Nothing  -> (Just $ Insert wid $ singleton point cp, ())
+                    Nothing  -> (Nothing, ())
                     Just _   -> (Just $ Adjust wid $ PutCheckpoint point cp, ())
 
+    -- Insert genesis checkpoint into the DBVar.
+    -- Throws an internal error if the checkpoint is not actually at genesis.
+    let insertCheckpointGenesis wid cp
+            | W.isGenesisBlockHeader header =
+                updateDBVar checkpointsVar $ Insert wid $ fromGenesis cp
+            | otherwise =
+                throwIO $ ErrInitializeNotGenesis wid header
```

Comment:
thinking if exposing `ErrInitializeNotGenesis` as `pure $ Left ...`. we could have, then no `ErrNoSuchWallet` here
```
    , putCheckpoint
        :: WalletId
        -> Wallet s
        -> ExceptT ErrNoSuchWallet stm ()
```
but ErrPutCheckpoint which is sum type witth error no wallet and initialize not genesis?

Reply:
I opted to go for an exception here, because `ErrInitializeNotGenesis` is a programming error (it's the responsibility of the caller of `initializeWallet` to make sure that only the genesis checkpoint is passed), whereas `ErrNoSuchWallet` may be a user error (the user specified a wallet ID that is not in the database).